### PR TITLE
[Doc] Update docstring for ndrange

### PR DESCRIPTION
**Pull request by neozhaoliang**

### CODE REVIEW

File: python/taichi/lang/impl.py [hunk 1]
```diff
@@ -786,15 +826,24 @@ def static(x, *xs):
 
 @taichi_scope
 def grouped(x):
-    """Groups a list of independent loop indices into a :func:`~taichi.lang.matrix.Vector`.
+    """Groups the indices in the iterator returned by `ndrange()` into a 1-D vector.
+
+    This is often used when you want to iterator over all indices returned by `ndrange()`
```
Comment (strongoier) [inline python/taichi/lang/impl.py:831]:
```suggestion
    This is often used when you want to iterate over all indices returned by `ndrange()`
```

File: python/taichi/lang/impl.py [hunk 2]
```diff
@@ -505,13 +520,18 @@ def field(dtype, shape=None, name="", offset=None, needs_grad=False):
 
     Args:
         dtype (DataType): data type of the field.
-        shape (Union[int, tuple[int]], optional): shape of the field
-        name (str, optional): name of the field
-        offset (Union[int, tuple[int]], optional): offset of the field domain
+
+        shape (Union[int, tuple[int]], optional): shape of the field.
+
```
Comment (strongoier) [inline python/taichi/lang/impl.py:525]: Why empty lines?

File: python/taichi/lang/_ndrange.py [hunk 3]
```diff
@@ -43,6 +43,85 @@ def grouped(self):
...
 
 def ndrange(*args):
+    """Return an immutable iterator object for looping over multi-dimensional indices.
+
+    This returned set of multi-dimensional indices is the direct product (in the set-theory sense)
+    of n groups of integers, where n equals the number of arguments in the input list, and looks like
+
+    range(x1, y1) x range(x2, y2) x ... x range(xn, yn)
+
+    The k-th argument correspondes to the k-th `range()` factor in the above product, and each
```
Comment (strongoier) [inline python/taichi/lang/_ndrange.py:53]:
```suggestion
    The k-th argument corresponds to the k-th `range()` factor in the above product, and each
```

File: python/taichi/lang/impl.py (hunk 2 above)
Comment (neozhaoliang) [reply python/taichi/lang/impl.py:525]: I thought adding empty lines could improve readability :P:

File: python/taichi/tools/video.py [hunk 4]
```diff
@@ -54,6 +54,30 @@ def mp4_to_gif(input_fn, output_fn, framerate):
...
+        output_dir (str): directory to save the frames.
+        width, height (int): resolution of the video.
+        post_processor (object): any object that implements the `process(img)` method, which accepts
+            an image as a numpy.ndarray and returns the process image.
+        framerate (int): frame rate of the video.
+        automatic_build (bool): automatically generate the resulting video or not.
+
+    Example::
+
+        >>> video_manager = ti.VideoManager(output_dir="./output", framerate=24, automatic_build=False)
```
[the remark this answers is not in the document]
Comment (neozhaoliang) [reply python/taichi/tools/video.py]: corrected

File: python/taichi/lang/snode.py [hunk 5]
```diff
@@ -399,12 +436,13 @@ def get_addr(f, indices):
     Currently, this function can only be called inside a taichi kernel.
 
     Args:
-        f (Union[ti.field, ti.Vector.field, ti.Matrix.field]): Input taichi field for memory address query.
-        indices (Union[int, ti.Vector()]): The specified field indices of the query.
+        f (Union[:class:`~taichi.field`, :class:`~taichi.Vector.field`, \
```
Comment (strongoier) [inline python/taichi/lang/snode.py:439]: `taichi.field` is not a class - `taichi.Field` is. This should be corrected to get the hyperlink work.

File: python/taichi/lang/matrix.py [hunk 6]
```diff
@@ -1027,15 +1061,24 @@ def outer_product(self, other):
...
+        dt (:mod:`~taichi.types.primitive_types`): data type of the vector.
 
     Returns:
-        :class:`~taichi.lang.matrix.Matrix`: A Vector instance (1-D :class:`~taichi.lang.matrix.Matrix`).
-
+        :class:`~taichi.Matrix`: A vector instance.
+
+    Example::
+        >>> u = ti.Vector([1, 2])
+        >>> print(u.m, u.n)  # verify a vector is a matrix of shape (n ,1)
```
Comment (strongoier) [inline python/taichi/lang/matrix.py:1077]:
```suggestion
        >>> print(u.m, u.n)  # verify a vector is a matrix of shape (n, 1)
```

File: python/taichi/lang/snode.py (hunk 5 above)
Comment (strongoier) [reply python/taichi/lang/snode.py]: `taichi.Matrix.field` should become `taichi.MatrixField` accordingly.

Comment (neozhaoliang) [reply python/taichi/lang/snode.py]: Corrected.

File: python/taichi/lang/snode.py [hunk 7]
```diff
@@ -366,31 +361,73 @@ def rescale_index(a, b, I):
...
+def deactivate(node, indices):
+    """Explicitly deactivate a cell of `node` at location `indices`.
 
+    After deactivation,  the Taichi runtime automatically recycles
+    and zero-fills memory of the deactivated cell.
 
-def deactivate(l, indices):
+    Args:
+        node (:class:`~taichi.SNode`): Must be a pointer, hash or bitmasked node.
+        indices (Union[int, :class:`~taichi.Vector`]): the indices to deactivate.
```
Comment (mzmzm) [inline python/taichi/lang/snode.py:413]: `list` is supported in `indices` too.

Comment (neozhaoliang) [reply python/taichi/lang/snode.py:413]: corrected.

File: python/taichi/lang/snode.py (hunk 5 above)
Comment (strongoier) [reply python/taichi/lang/snode.py]: `taichi.Vector.field` is also `taichi.MatrixField`.

Comment (neozhaoliang) [reply python/taichi/lang/snode.py]: corrected

Comment (strongoier) [reply python/taichi/lang/snode.py]: As the two `taichi.MatrixField`s are the same, shall we not duplicate them?

Comment (neozhaoliang) [reply python/taichi/lang/snode.py]: you are right, I'm sorry for this careless mistake.

File: python/taichi/lang/snode.py [hunk 8]
```diff
@@ -330,22 +330,17 @@ def _physical_index_position(self):
...
-    Returns
-    -------
-    Ib: ti.Vector()
-        rescaled grouped loop index
+    """Rescales the index 'I' of field (or SNode) 'a' to match the shape of SNode 'b'.
 
+    Args:
+
+        a, b (Union[:class:`~taichi.Field`, :class:`~taichi.MatrixField`, \
+            :class:`~taichi.MatrixField`): Input taichi fields or snodes.
```
Comment (strongoier) [inline python/taichi/lang/snode.py:338]: Same here.

Comment (neozhaoliang) [reply python/taichi/lang/snode.py]: corrected.

**Review by strongoier — APPROVED** [review]
LGTM!

**Pull request merged**